saw-python/README.md

```diff
@@ -0,0 +1,191 @@
+
+### Execute Functions
+The command `self.execute_func(inputs)` will symbolically execute the function. There should be exactly as many comma separated inputs as there are 
+in the C function. As usual with SAW, one can only place preconditions before this command and postconditions after this command 
+(there are no preconditions or postconditions in this contract; we'll discuss those later).
+
+### Return Statements
+The command `self.returns_f(string)` asserts the current function returns the Cryptol term parsed from a Python String. To use Python variables in scope within the string use 
+`{variable_name}`. For example,
+
+|`self.`|`returns_f`|`("{bits} >>> {shift}")`|
+|-------|-----------|---------------------|
+|In this contract| assert the current function returns the Cryptol term | circular shift the bit array `xs` by `shift`|
+
+Sometimes we don't want to return a Cryptol Term. In this case we can just use `returns(someSetupValue)`. The specification function of a Contract must
+**always** have a `self.returns(someSetupValue)` or `self.returns_f(string)` statement. If the function returns `void` one can use `self.returns(void)`.
+
+### Terms from Cryptol
+The `CryptolTerm` class is a subclass of `SetupVal`, so you can use them anywhere you'd use a `SetupVal` in SAW (in non-Python SAW they are disjoint types).
+
+The command `cry(string)` converts a Python string into a Cryptol Term that can be used in SAW. The `_f` indicates one can pass in Python local variables into the strings by surrounding
+the variable with braces as we did in `{bits} >>> {shift}`. In fact, `returns_f` is just syntactic sugar for `returns(cry_f(string))`.
+
+Braces are sometimes used in Cryptol to assign type parameters or record types. To have the parser parse a literal brace in a Cryptol string repeat the `{` or `}` symbols.
+For example, consider the following made up SAW line:
+
+```python
+  self.returns_f("{{a = take`{{5}} {blah}, b = take`{{{N}}} blah }} == foo `blah")
+```
+
+We parse the Cryptol string as a boolean assertion between records
+
+```cryptol
+{a = take`{5} 23, b = take`{2} 23} == foo `blah
+```
+
+where `blah` is a local Python variable equal to `23`, `N` is a local Python variable equal to `2`, `foo` some function returning a record, and `blah` some 
+Cryptol type in the specification loaded.
+
+### Preconditions and Postconditions
+
+The [C specification](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf) has the following to say about bit shifts:
+
+> If the value of the right operand is negative or is
+> greater than or equal to the width of the promoted left operand, the behavior is undefined.
+
+Due to this, we should probably add a precondition to the previous code:
+
+```python
+class RCS_Contract(Contract):
+  def specification(self):
+    xs    = self.fresh_var(i32, "xs") 
+    shift = self.fresh_var(i32, "shift")
+    
+    self.precondition_f("{shift} <= 32")

```

Comment:
This precondition isn't enough because `bits >> (sizeof(bits)*8 - shift)` exhibits undefined behavior when `shift` is `0` (`sizeof(bits)*8 - shift` evaluates to `32`).

Reply:
It's kind of neat actually to see SAW find these errors. See the lab being developed for Python SAW here: https://github.com/weaversa/cryptol-course/blob/saw-python-api/labs/SAW/SAW.md
Roughly lines 300-400.